### PR TITLE
Better handling of new/sparse Growth and Finance topics

### PR DESCRIPTION
**Pull request by belucid**
https://trello.com/c/ro7Miff6

Review with latest mainline. PULL!

Test on both big web and mobile (where applicable):

- [x] Create a new company with `/create-company`
- [x] Look at the front/collapsed of the Growth placeholder topic, is it blank? Good.
- [x] Look at the expanded/back of the Growth placeholder topic, is there a helpful new message? Hot damn.
- [x] Can you click the pencil and edit the Growth placeholder topic without getting an error? Mazeltov!
- [x] Add a new metric, but don't add any data. Front/collapsed and back/expanded still blank. Nice!
- [x] Add just some sparse 0 data to  the new metric, but don't add any non-0 data. Front/collapsed and back/expanded still blank. Very nice!
- [x] Now add some sparse non-0 data to the new metric. Front/collapsed and expanded/back of the topic have content now? Great gravy!
- [x] Look at the front/collapsed of the Finance placeholder topic, is it blank? Splendid.
- [x] Look at the expanded/back of the Finance placeholder topic, is there a helpful new message? Don't hide that in a bushel basket!
- [x] Add just some sparse 0 data to the finances, but don't add any non-0 data. Front/collapsed and back/expanded still blank. Hum diggity!
- [x] Now add some sparse non-0 data to the new metric. Front/collapsed and expanded/back of the topic have content now? Woza!
- [x] Repeat the same above, but with a new Growth card on an existing company
- [x] Repeat the same above, but with a new Finance card on an existing company
- [x] Sing Purple Rain, acapella and falsetto, in nothing but purple undies, in the rain

### CODE REVIEW

**Comment by belucid** [comment]
@martinklepsch you're up to review this.

**Comment by martinklepsch** [comment]
noted. Somehow the assigning earlier slipped through my github notifications. Will review later today

**Comment by martinklepsch** [comment]
@belucid When attempting to create a company I get an error in the console: 
```
Cannot dispatch in the middle of a dispatch.
```
After fixing that the growth card that is part of the new company has a lot of buttons on it, i.e. it is not blank. Am I doing something wrong?

**Comment by belucid** [comment]
@martinklepsch this is resolved in another branch: "When attempting to create a company I get an error in the console" so don't worry about it here

**Comment by belucid** [comment]
@martinklepsch you aren't running the latest API if you're seeing a growth topic that's not blank

**Comment by martinklepsch** [comment]
>  Can you click the pencil and edit the Growth placeholder topic without getting an error? Mazeltov!

While this works the interface itself seems pretty confusing. When adding a new metric the primary action is very hard to spot.

screenshot 2016-05-02 20 24 23

**Comment by martinklepsch** [comment]
> Now add some sparse non-0 data to the new metric. Front/collapsed and expanded/back of the topic have content now? Great gravy!

This works but when moving the cursor away from the chart it always focuses the last data point. I think we fixed this before so that the last focused remains focused?

**Comment by martinklepsch** [comment]
> Information on finances is not yet available.
Are these the "helpful messages"?

**Comment by martinklepsch** [comment]
After adding a metric to the Growth card the Growth card looks like this:
screenshot 2016-05-02 20 37 34

**Comment by belucid** [comment]
@martinklepsch I couldn't agree more that the Growth interface is confusing. It's not a UI from one of our designers, it's just from us (developers) as something that could functionally work to show the topic card, We expect to get a new design on this from Mark this week.

**Comment by belucid** [comment]
@martinklepsch No, I'm not certain those messages are helpful enough. It's an interesting copy writing challenge. I think it's more helpful than blank. You have the case of a r/o or r/w user (which can get different messages) and the tougher case of the r/w user that should do something about this (a founder, an exec., versus a r/w user that shouldn't do anything about this (a low level employee), and we can't tell these 2 apart and show them different messages.

I'm not certain what would be a more helpful message to show a r/o user, and what we can say as 1 thing that's helpful to both types of r/w users. That why we ended up with what's there, that's maybe not very helpful...

**Comment by belucid** [comment]
@martinklepsch the charts got completely reworked in another branch to the new design so I'm not concerned about the focus issue here.

**Comment by belucid** [comment]
@martinklepsch the front of these topic cards is getting a completely new design in the card @bago2k4 is working on right now so I'm not concerned about the cut off metric pillbox here.

**Comment by martinklepsch** [comment]
I wasn't complaining about the "helpful messages" just wanted to make sure I'm seeing what I'm supposed to see. :)

**Comment by belucid** [comment]
Ah... that's them. Again... not sure "helpful" was the best description. Hopefully we'll dream up better copy at some point.

**Comment by martinklepsch** [comment]
>  Repeat the same above, but with a new Growth card on an existing company
This worked as expected after adding a Growth section to Yoyo. However even with "Measured as a number" the values are formatted as dollar-values.

**Comment by martinklepsch** [comment]
> Repeat the same above, but with a new Finance card on an existing company

Worked as well. Again: Dollar values were rendered like `$$100`. Also tabbing through the input fields works well but shift-tab doesn't move to the previous field.

File: src/open_company_web/lib/utils.cljs [hunk 1]
```diff
@@ -504,20 +504,37 @@
     (> current-month 9)
     10))
 
+(defn non-zero-number?
+  [number]
+  ; Seems a little convoluted, but need it to short circuit if it's not a number
+  (not (or (not (number? number)) (zero? number))))
```
Comment (martinklepsch) [inline src/open_company_web/lib/utils.cljs]: How about this?
```clojure
(and (number? number) 
     (not (zero? number))
```

File: src/open_company_web/lib/utils.cljs [hunk 2]
```diff
@@ -504,20 +504,37 @@
...
+  [data]
+  (not (some #(or (non-zero-number? (:cash %))
+                  (non-zero-number? (:revenue %))
+                  (non-zero-number? (:costs %))) data)))
+
+(defn no-growth-data?
+  "Return `true` if the passed finances data has no data, otherwise return `false`."
+  [data]
+  (not (some #(non-zero-number? (:value %)) (vals data))))
+
```
Comment (martinklepsch) [inline src/open_company_web/lib/utils.cljs]:
```clojure
cljs.user=> (not (some true? [false true false]))
false
cljs.user=> (not-any? true? [false true false])
false
```

File: src/open_company_web/components/topic_overlay_edit.cljs [hunk 3]
```diff
@@ -210,7 +211,7 @@
     {:data fixed-growth-data}))
 
 (defn data-to-save [owner topic]
- (let [topic-kw (keyword topic)
+  (let [topic-kw (keyword topic)
```
Comment (martinklepsch) [inline src/open_company_web/components/topic_overlay_edit.cljs]: Indenting here but not indenting the following let-exprs doesn't seem consistent.

File: src/open_company_web/components/topic_overlay_edit.cljs [hunk 4]
```diff
@@ -107,7 +107,7 @@
           focus-metric (or growth-metric-focus (:slug (first all-metrics)))]
       {:growth-focus (or focus-metric growth-utils/new-metric-slug-placeholder)
        :growth-metadata-editing false
-       :growth-new-metric false
+       :growth-new-metric (zero? (count all-metrics))
```
Comment (martinklepsch) [inline src/open_company_web/components/topic_overlay_edit.cljs]:
```clojure
cljs.user=> (zero? (count []))
true
cljs.user=> (empty? [])
true
```

File: src/open_company_web/lib/utils.cljs [hunk 5]
```diff
@@ -504,20 +504,37 @@
...
-      "monthly"
-      (str year "-" (add-zero month))
       "weekly"
       (let [day-of-week (cljs-time/day-of-week now)
             to-monday (dec day-of-week)
             monday-date (cljs-time/minus now (cljs-time/days to-monday))]
-        (str (cljs-time/year monday-date) "-" (add-zero (cljs-time/month monday-date)) "-" (add-zero (cljs-time/day monday-date)))))))
+        (str (cljs-time/year monday-date) "-" (add-zero (cljs-time/month monday-date)) "-" (add-zero (cljs-time/day monday-date))))
+      ;; Default tp monthly
+      (str year "-" (add-zero month)))))
```
Comment (martinklepsch) [inline src/open_company_web/lib/utils.cljs]: Not a fan of formatting things by hand like this when we have things like `goog.date.Date` and `goog.i18n.DateTimeFormat`. Even cljs-time which is used here has formatting capabilities: https://github.com/andrewmcveigh/cljs-time#cljs-timeformat

Comment (belucid) [inline src/open_company_web/lib/utils.cljs]: @bago2k4 this is yours, consider switching over to one of the format functions in another branch you're working on.

File: src/open_company_web/lib/utils.cljs (hunk 1 above)
Comment (belucid) [inline src/open_company_web/lib/utils.cljs]: Switched.

File: src/open_company_web/lib/utils.cljs (hunk 2 above)
Comment (belucid) [inline src/open_company_web/lib/utils.cljs]: Switched.

File: src/open_company_web/components/topic_overlay_edit.cljs (hunk 4 above)
Comment (belucid) [inline src/open_company_web/components/topic_overlay_edit.cljs]: switched

File: src/open_company_web/components/topic_overlay_edit.cljs (hunk 3 above)
Comment (belucid) [inline src/open_company_web/components/topic_overlay_edit.cljs]: Fixed.

**Pull request merged**